File: public/code/memecoin__volume_surge.py
"""
Talons Strategies - Volume Surge Detection (Memecoin)

STRATEJI:
- Klasik volume_avg vs volume_now karşılaştırması
- 5dk/15dk/1h farklı timeframe'lerde surge
- Buy pressure (buy volume > sell volume) kontrolü
- Price tepkisi kontrol (volume up + price up = long, volume up + price down = avoid)

INPUT: list[Candle] (5dk OHLCV) - en az 50 mum
"""
import sys, os
sys.path.append(os.path.join(os.path.dirname(__file__), "..", ".."))
from common import make_signal, atr, sma, validate_ohlcv


def check_signal(candles: list, cfg: dict = None) -> dict:
    cfg = cfg or {}
    surge_mult = cfg.get("surge_multiplier", 3.0)
    lookback   = cfg.get("volume_lookback", 20)
    atr_mult   = cfg.get("sl_atr_mult", 1.5)
    rr1, rr2, rr3 = cfg.get("rr_targets", (2.0, 3.5, 5.0))

    if not validate_ohlcv(candles, lookback + 5):
        return make_signal("neutral", 0, 0, [], 0, "Not enough candles")

    vols   = [c["volume"] for c in candles]
    closes = [c["close"] for c in candles]
    opens  = [c["open"] for c in candles]
    avg_v  = sum(vols[-lookback-1:-1]) / lookback
    now_v  = vols[-1]

    if now_v < avg_v * surge_mult:
        return make_signal("neutral", closes[-1], 0, [], 0, f"No surge ({now_v/max(avg_v,1):.2f}x)")

    bull = closes[-1] > opens[-1]
    a = atr(candles, 14)
    price = closes[-1]

    if not bull:
        return make_signal("neutral", price, 0, [], 0, "Volume surge on red candle - avoid")

    sl = price - a * atr_mult
    risk = price - sl
    conf = min(0.5 + (now_v/avg_v - surge_mult) * 0.05 + 0.2, 0.95)

    return make_signal(
        "long", price, sl,
        [price + risk * rr1, price + risk * rr2, price + risk * rr3],
        conf,
        f"Volume surge {now_v/avg_v:.2f}x avg, bullish candle",
        metadata={"surge_ratio": now_v/avg_v},
    )

File: public/code/memecoin__volume_surge.ts
// Talons Strategies - Volume Surge Detection (Memecoin)
import { makeSignal, Signal, Candle, atr, validateOhlcv } from "../../common";

export interface VolSurgeConfig {
  surge_multiplier?: number;
  volume_lookback?: number;
  sl_atr_mult?: number;
  rr_targets?: [number, number, number];
}

export function checkSignal(candles: Candle[], cfg: VolSurgeConfig = {}): Signal {
  const mult     = cfg.surge_multiplier ?? 3.0;
  const lookback = cfg.volume_lookback ?? 20;
  const slMul    = cfg.sl_atr_mult ?? 1.5;
  const [rr1, rr2, rr3] = cfg.rr_targets ?? [2.0, 3.5, 5.0];

  if (!validateOhlcv(candles, lookback + 5))
    return makeSignal("neutral", 0, 0, [], 0, "Not enough candles");

  const vols   = candles.map(c => c.volume);
  const closes = candles.map(c => c.close);
  const opens  = candles.map(c => c.open);

  const window = vols.slice(-lookback-1, -1);
  const avgV = window.reduce((a,b)=>a+b,0) / window.length;
  const nowV = vols[vols.length-1];

  if (nowV < avgV * mult)
    return makeSignal("neutral", closes[closes.length-1], 0, [], 0,
                     `No surge (${(nowV/Math.max(avgV,1)).toFixed(2)}x)`);

  const bull = closes[closes.length-1] > opens[opens.length-1];
  const atrArr = atr(candles, 14);
  const a = atrArr[atrArr.length - 1];
  const price = closes[closes.length-1];

  if (!bull)
    return makeSignal("neutral", price, 0, [], 0, "Volume surge on red - avoid");

  const sl = price - a * slMul;
  const risk = price - sl;
  const conf = Math.min(0.5 + (nowV/avgV - mult) * 0.05 + 0.2, 0.95);

  return makeSignal("long", price, sl,
                    [price + risk*rr1, price + risk*rr2, price + risk*rr3],
                    conf,
                    `Volume surge ${(nowV/avgV).toFixed(2)}x avg, bullish candle`,
                    { surge_ratio: nowV/avgV });
}
